doc/configuration.md
# Configuration

Listed below are all configuration options.

* `adapter` - useful only for the gem developers to switch to a new
  adapter. Default and the only available value is `aws_sdk_v3`
* `namespace` - prefix for table names, default is
  `dynamoid_#{application_name}_#{environment}` for Rails application
  and `dynamoid` otherwise
* `logger` - by default it's a `Rails.logger` in Rails application and
  `stdout` otherwise. You can disable logging by setting `nil` or
  `false` values. Set `true` value to use defaults
* `access_key` - DynamoDB custom access key for AWS credentials, override global
  AWS credentials if they're present
* `secret_key` - DynamoDB custom secret key for AWS credentials, override global
  AWS credentials if they're present
* `credentials` - DynamoDB custom pre-configured credentials, override global
  AWS credentials if they're present
* `region` - DynamoDB custom credentials for AWS, override global AWS
  credentials if they're present
* `batch_size` - when you try to load multiple items at once with
* `batch_get_item` call Dynamoid loads them not with one api call but
  in chunks. Default is 100 items
* `capacity_mode` - used at a table creation and means whether a table
  read/write capacity mode will be on-demand or provisioned. Allowed
  values are `:on_demand` and `:provisioned`. Default value is `nil` which
  means provisioned mode will be used.
* `read_capacity` - is used during table or index creation. Default is 100
  (units)
* `write_capacity` - is used during table or index creation. Default is 20
  (units)
* `warn_on_scan` - log warnings when scan table. Default is `true`
* `error_on_scan` - raises an error when scan table. Default is `false`
* `endpoint` - if provided, it communicates with the DynamoDB listening
  at the endpoint. This is useful for testing with
  [DynamoDB Local](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Tools.DynamoDBLocal.html)
* `identity_map` - ensures that each object gets loaded only once by
  keeping every loaded object in a map. Looks up objects using the map
  when referring to them. Isn't thread safe. Default is `false`.
  `Use Dynamoid::Middleware::IdentityMap` to clear identity map for each HTTP request
* `timestamps` - by default Dynamoid sets `created_at` and `updated_at`
  fields at model creation and updating. You can disable this
  behavior by setting `false` value
* `sync_retry_max_times` - when Dynamoid creates or deletes table
  synchronously it checks for completion specified times. Default is 60
  (times). It's a bit over 2 minutes by default
* `sync_retry_wait_seconds` - time to wait between retries. Default is 2
  (seconds)
* `convert_big_decimal` - if `true` then Dynamoid converts numbers
  stored in `Hash` in `raw` field to float. Default is `false`
* `store_attribute_with_nil_value` - if `true` Dynamoid keeps attribute
  with `nil` value in a document. Otherwise Dynamoid removes it while
  saving a document. Default is `nil` which equals behaviour with `false`
  value.
* `models_dir` - `dynamoid:create_tables` rake task loads DynamoDB
  models from this directory. Default is `./app/models`.
* `application_timezone` - Dynamoid converts all `datetime` fields to
  specified time zone when loads data from the storage.
  Acceptable values - `:utc`, `:local` (to use system time zone) and
  time zone name e.g. `Eastern Time (US & Canada)`. Default is `utc`
* `dynamodb_timezone` - When a datetime field is stored in string format
  Dynamoid converts it to specified time zone when saves a value to the
  storage. Acceptable values - `:utc`, `:local` (to use system time
  zone) and time zone name e.g. `Eastern Time (US & Canada)`. Default is
  `utc`
* `store_datetime_as_string` - if `true` then Dynamoid stores :datetime
  fields in ISO 8601 string format. Default is `false`
* `store_date_as_string` - if `true` then Dynamoid stores :date fields
  in ISO 8601 string format. Default is `false`
* `store_empty_string_as_nil` - store attribute's empty String value as NULL. Default is `true`
* `store_boolean_as_native` - if `true` Dynamoid stores boolean fields
  as native DynamoDB boolean values. Otherwise boolean fields are stored
  as string values `'t'` and `'f'`. Default is `true`
* `store_binary_as_native` - if `true` Dynamoid stores binary fields
  as native DynamoDB binary values. Otherwise binary fields are stored
  as Base64 encoded string values. Default is `false`
* `backoff` - is a hash: key is a backoff strategy (symbol), value is
  parameters for the strategy. Is used in batch operations. Default id
  `nil`
* `backoff_strategies`: is a hash and contains all available strategies.
  Default is `{ constant: ..., exponential: ...}`
* `log_formatter`: overrides default AWS SDK formatter. There are
  several canned formatters: `Aws::Log::Formatter.default`,
  `Aws::Log::Formatter.colored` and `Aws::Log::Formatter.short`. Please
  look into `Aws::Log::Formatter` AWS SDK documentation in order to
  provide own formatter.
* `http_continue_timeout`: The number of seconds to wait for a
  100-continue HTTP response before sending the request body. Default
  option value is `nil`. If not specified effected value is `1`
* `http_idle_timeout`: The number of seconds an HTTP connection is
  allowed to sit idle before it is considered stale. Default option
  value is `nil`. If not specified effected value is `5`
* `http_open_timeout`: The number of seconds to wait when opening a HTTP
  session. Default option value is `nil`. If not specified effected
  value is `15`
* `http_read_timeout`:The number of seconds to wait for HTTP response
  data. Default option value is `nil`. If not specified effected value
  is `60`
* `http_proxy`: A proxy to send requests through. Formatted like
  `'http://proxy.com:1234'`. Default option value is `nil`
* `create_table_on_save`: if `true` then Dynamoid creates a
  corresponding table in DynamoDB at model persisting if the table
  doesn't exist yet. Default is `true`
* `use_yaml_unsafe_load`: allows deserializing arbitrary classes in
  YAML-serialized fields. Default is `true` on Ruby < 3.1, and `false`
  starting from 3.1
* `yaml_permitted_classes`: list of additional classes permitted when
  deserializing YAML-serialized fields. Default is
  `[Symbol, Set, Date, Time, DateTime]`


doc/test_environment.md
# Test Environment

In test environment you will most likely want to clean the database
between test runs to keep tests completely isolated. This can be
achieved like so

```ruby
module DynamoidReset
  def self.all
    Dynamoid.adapter.list_tables.each do |table|
      # Only delete tables in our namespace
      if table =~ /^#{Dynamoid::Config.namespace}/
        Dynamoid.adapter.delete_table(table)
      end
    end
    Dynamoid.adapter.tables.clear
    # Recreate all tables to avoid unexpected errors
    Dynamoid.included_models.each { |m| m.create_table(sync: true) }
  end
end

# Reduce noise in test output
Dynamoid.logger.level = Logger::FATAL
```

If you're using RSpec you can invoke the above like so:

```ruby
RSpec.configure do |config|
  config.before(:each) do
    DynamoidReset.all
  end
end
```

In addition, the first test for each model may fail if the relevant models are not included in `included_models`. This can be fixed by adding this line before the `DynamoidReset` module:
```ruby
Dir[File.join(Dynamoid::Config.models_dir, '**/*.rb')].sort.each { |file| require file }
```
Note that this will require _all_ models in your models folder - you can also explicitly require only certain models if you would prefer to.

In Rails, you may also want to ensure you do not delete non-test data
accidentally by adding the following to your test environment setup:

```ruby
raise "Tests should be run in 'test' environment only" if Rails.env != 'test'

Dynamoid.configure do |config|
  config.namespace = "#{Rails.application.railtie_name}_#{Rails.env}"
end
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Fixed
### Added
### Changed
### Removed

## 3.14.0 / 2026-07-26

### Fixed
* [#1006](https://github.com/Dynamoid/dynamoid/pull/1006) Fix non-transactional modifying methods to:
    * preserve explicitly specified `lock_version` values
    * preserve an already persisted `nil` value of `lock_version` during model persistence
    * skip locking when `lock_version` is `nil`
* [#1007](https://github.com/Dynamoid/dynamoid/pull/1007) Fix transactional `#update_fields` and `#save` methods to support DynamoDB reserved words as partition and sort key names
* [#1007](https://github.com/Dynamoid/dynamoid/pull/1007) Fix non-transactional `#inc` method to
    * raise `UnknownAttribute` when passed an unknown field name
    * raise `MissingHashKey` or `MissingRangeKey` when the partition or sort key is `nil` or omitted
* [#1007](https://github.com/Dynamoid/dynamoid/pull/1007) Fix transactional `#delete`, `#update_fields`, and `#upsert` methods to type-cast partition and sort key values
* [#1010](https://github.com/Dynamoid/dynamoid/pull/1010) Fix transactional `#delete(model)` to run `after_commit` and `after_rollback` callbacks
### Added
* [#1005](https://github.com/Dynamoid/dynamoid/pull/1005) Introduce public methods to start transactions (replace the removed `Dynamoid::TransactionRead` class):
    * `Dynamoid::Document.transaction`
    * `Dynamoid::Document.transaction.writing`
    * `Dynamoid::Document.transaction.reading`
* [#1007](https://github.com/Dynamoid/dynamoid/pull/1007) Add transactional `#inc` method
* [#1010](https://github.com/Dynamoid/dynamoid/pull/1010) Add missing transactional methods:
    * `#increment!`, `#decrement!`
    * `#touch`
    * `#update_attribute`, `#update_attribute!`
    * `#import`
### Changed
* [#1006](https://github.com/Dynamoid/dynamoid/pull/1006) Implement optimistic locking for transactional methods using the `lock_version` field. The following methods now support it:
    * `#save`, `#save!`
    * `#create`, `#create!`
    * `#update_attributes`, `#update_attributes!`
    * `#update_attribute`
    * `#delete`
    * `#destroy`, `#destroy!`
### Removed
* [#1005](https://github.com/Dynamoid/dynamoid/pull/1005) Remove the `Dynamoid::TransactionRead` class (which was part of the experimental API)

## 3.13.1 / 2026-05-05

### Fixed
* [#1003](https://github.com/Dynamoid/dynamoid/pull/1003) Fix detecting dirty changes for `:map` fields

## 3.13.0 / 2026-03-29

### Fixed
* [#944](https://github.com/Dynamoid/dynamoid/pull/944) Fix `#delete` and `#destroy` methods and set `#destroyed?` properly when operations fail
* [#987](https://github.com/Dynamoid/dynamoid/pull/987) Fix checking that a primary key is given in transactional methods `#save` and `#destroy`
### Added
* [#941](https://github.com/Dynamoid/dynamoid/pull/941) Support table ARN and add option `:arn` for the `table` method to specify a table belonged to specific AWS account
* [#943](https://github.com/Dynamoid/dynamoid/pull/943) Implement `delete` class method
* [#945](https://github.com/Dynamoid/dynamoid/pull/945) Implement `#update_attribute!` method
* [#947](https://github.com/Dynamoid/dynamoid/pull/947) Allow skipping default model fields generation and add option `:skip_generating_fields` for the `table` method to specify field names
* [#988](https://github.com/Dynamoid/dynamoid/pull/988) Add Ruby 4.0 and Rails 8.1 in CI

## 3.12.0 / 2025-08-23

### Fixed
* [#849](https://github.com/Dynamoid/dynamoid/pull/849) Fixed saving a field of custom type when it implements both `.dynamoid_dump()` and `#dynamoid_dump` method and use the former one.
A proper behaviour is to use adapter's .dynamoid_dump method instead of a value's #dynamoid_dump method.
* [#851](https://github.com/Dynamoid/dynamoid/pull/851) Fixed Dirty API and don't require custom types to implement `#==` method. Add field option `:comparable`.
* [#913](https://github.com/Dynamoid/dynamoid/pull/913) Fixed saving partition keys of Dynamoid-specific types (e.g. `date` or `datetime`) and always convert them into a proper DynamoDB type
* [#917](https://github.com/Dynamoid/dynamoid/pull/917) Fixed transactional write operations when primary key is of non-native DynamoDB type
* [#927](https://github.com/Dynamoid/dynamoid/pull/927) Multiple fixes in transactional and non-transactional write methods and finders:
  * Changed non-transactional method `#find` and raise `MissingHashKey` when a given partition key is `nil`
  * Fixed non-transactional method `#save!` and raise `RecordNotSaved` when a callback throws `:abort` and terminates the process
  * Check whether partition and sort key are specified in the non-transactional persistence methods and raise a proper exception (`MissingHashKey` or `MissingRangeKey`)
  * Fixed methods `#inc`, `#increment!` and `#decrement!`, `#update` and `#update!` to not create a new item when an item with specified primary key is already deleted
  * Fixed method `#update_attribute` to not raise `StaleObjectError` when an item with specified primary key is already deleted
  * Fixed method `#where` with condition on a `:serialized` field to not raise an exception
### Added
* [#910](https://github.com/Dynamoid/dynamoid/pull/910) Added `key_type` option to the `table` method to change type of a partition key field (@ogidow)
* [#916](https://github.com/Dynamoid/dynamoid/pull/916) Added new `error_on_scan` config option to raises an error and prevent table scans (DynamoDB's `Scan` operation) (@aaronalberg)
* [#926](https://github.com/Dynamoid/dynamoid/pull/926) Implemented read transactions (using DynamoDB's `TransactGetItems` operation)
### Changed

lib/dynamoid/config.rb
# frozen_string_literal: true

require 'uri'
require 'logger'
require 'date'
require 'set'
require 'dynamoid/config/options'
require 'dynamoid/config/backoff_strategies/constant_backoff'
require 'dynamoid/config/backoff_strategies/exponential_backoff'

module Dynamoid
  # Contains all the basic configuration information required for Dynamoid: both sensible defaults and required fields.
  # @private
  module Config
    # @since 3.3.1
    DEFAULT_NAMESPACE = if defined?(Rails)
                          klass = Rails.application.class
                          app_name = Rails::VERSION::MAJOR >= 6 ? klass.module_parent_name : klass.parent_name
                          "dynamoid_#{app_name}_#{Rails.env}"
                        else
                          'dynamoid'
                        end

    extend self # rubocop:disable Style/ModuleFunction

    extend Options
    include ActiveModel::Observing if defined?(ActiveModel::Observing)

    # All the default options.
    option :adapter, default: 'aws_sdk_v3'
    option :namespace, default: DEFAULT_NAMESPACE
    option :access_key, default: nil
    option :secret_key, default: nil
    option :credentials, default: nil
    option :region, default: nil
    option :batch_size, default: 100
    option :capacity_mode, default: nil
    option :read_capacity, default: 100
    option :write_capacity, default: 20
    option :warn_on_scan, default: true
    option :error_on_scan, default: false
    option :endpoint, default: nil
    option :identity_map, default: false
    option :timestamps, default: true
    option :sync_retry_max_times, default: 60 # a bit over 2 minutes
    option :sync_retry_wait_seconds, default: 2
    option :convert_big_decimal, default: false
    option :store_attribute_with_nil_value, default: false # keep or ignore attribute with nil value at saving
    option :models_dir, default: './app/models' # perhaps you keep your dynamoid models in a different directory?
    option :application_timezone, default: :utc # available values - :utc, :local, time zone name like "Hawaii"
    option :dynamodb_timezone, default: :utc # available values - :utc, :local, time zone name like "Hawaii"
    option :store_datetime_as_string, default: false # store Time fields in ISO 8601 string format
    option :store_date_as_string, default: false # store Date fields in ISO 8601 string format
    option :store_empty_string_as_nil, default: true # store attribute's empty String value as null
    option :store_boolean_as_native, default: true
    option :store_binary_as_native, default: false
    option :backoff, default: nil # callable object to handle exceeding of table throughput limit
    option :backoff_strategies, default: {
      constant: BackoffStrategies::ConstantBackoff,
      exponential: BackoffStrategies::ExponentialBackoff
    }
    option :log_formatter, default: nil
    option :http_continue_timeout, default: nil # specify if you'd like to overwrite Aws Configure - default: 1
    option :http_idle_timeout, default: nil     #                                                  - default: 5
    option :http_open_timeout, default: nil     #                                                  - default: 15
    option :http_read_timeout, default: nil     #                                                  - default: 60
    option :http_proxy, default: nil
    option :create_table_on_save, default: true
    # Maintain backward compatibility: safe loading was unconditionally
    # introduced in Dynamoid 3.8.0 for Ruby >= 3.1.
    option :use_yaml_unsafe_load, default: Gem::Version.new(RUBY_VERSION) < Gem::Version.new('3.1.0')
    # Default list introduced in Dynamoid 3.8.0; kept for backward
    # compatibility until the next major version.
    option :yaml_permitted_classes, default: [Symbol, Set, Date, Time, DateTime]

    # The default logger for Dynamoid: either the Rails logger or just stdout.
    #
    # @since 0.2.0
    def default_logger
      defined?(Rails) && Rails.respond_to?(:logger) ? Rails.logger : ::Logger.new($stdout)
    end

    # Returns the assigned logger instance.
    #
    # @since 0.2.0
    def logger
      @logger ||= default_logger
    end

    # If you want to, set the logger manually to any output you'd like. Or pass false or nil to disable logging entirely.
    #
    # @since 0.2.0
    def logger=(logger)
      case logger
      when false, nil then @logger = ::Logger.new(nil)
      when true then @logger = default_logger
      else
        @logger = logger if logger.respond_to?(:info)
      end
    end

    def build_backoff
      if backoff.is_a?(Hash)
        name = backoff.keys[0]
        args = backoff.values[0]

        backoff_strategies[name].call(args)
      else
        backoff_strategies[backoff].call
      end
    end
  end
end
